src/providers/ogr/qgsogrfeatureiterator.cpp

```diff
@@ -282,20 +296,41 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
 
   gdal::ogr_feature_unique_ptr fet;
 
-  while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
-  {
-    if ( !readFeature( std::move( fet ), feature ) )
-      continue;
-
-    if ( !mFilterRect.isNull() && ( !feature.hasGeometry() || feature.geometry().isEmpty() ) )
-      continue;
+  // OSM layers (especially large ones) need the GDALDataset::GetNextFeature() call rather than OGRLayer::GetNextFeature()
+  // see more details here: https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite
 
-    // we have a feature, end this cycle
-    feature.setValid( true );
-    geometryToDestinationCrs( feature, mTransform );
-    return true;
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,2,0)
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) )
+  {
+    OGRLayerH nextFeatureBelongingLayer;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, &nextFeatureBelongingLayer, nullptr, nullptr, nullptr ) ), fet )
+    {
+      if ( checkFeature( fet, feature ) && QString::compare( QString::fromLatin1( OGR_L_GetName( nextFeatureBelongingLayer ) ), mSource->mLayerName ) == 0 )
```

Comment:
I think this check should be reversed -- the layer name comparison is much cheaper than the checkFeature, which requires reading the feature itself.

Reply:
And you can replace the layer name comparison by checking the equality of layer pointers directly.